### PR TITLE
[RF] Speed up getting weights from RooDataHist

### PR DESCRIPTION
**Pull request by guitargeek**
Retreiving weights from a RooDataHist can be sped up signifcantly if it
is assumed that parameters passed to `RooDataHist::weight` are aligned
with the histogram variables.

This commit introduces a `RooDataHist::weightFast` function that makes
this assumption. It is used by RooHistFunc and RooHistPdf.

This was benchmarked with the same hist factory example as in https://github.com/root-project/root/pull/7838:

* [without this PR, but with #7838](https://rembserj.web.cern.ch/rembserj/cgi-bin/igprof-navigator/hf_ParamHistFunc_1_new_2)
* [also with this PR](https://rembserj.web.cern.ch/rembserj/cgi-bin/igprof-navigator/hf_ParamHistFunc_1_new_3)

The effect is a 15 % speedup of the HistFactory example.

### CODE REVIEW

File: roofit/roofitcore/src/RooDataHist.cxx [hunk 1]
```diff
@@ -987,8 +987,10 @@ Int_t RooDataHist::getIndex(const RooAbsCollection& coord, Bool_t fast) const {
 /// as the internal coordinates. In this case, values are looked up only by index.
 std::size_t RooDataHist::calcTreeIndex(const RooAbsCollection& coords, bool fast) const
 {
-  // With fast, caller promises that layout of "coords" is identical to our internal "vars"
-  assert(!fast || _vars.size() == coords.size());
+  // Asserts commented out because the code is performance critical.
+
+  // With fast, caller promises that layout of "coords" is identical to our internal "vars".
+  // assert(!fast || _vars.size() == coords.size());
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooDataHist.cxx:993]: Check this out: [assert](https://en.cppreference.com/w/cpp/error/assert)
It's the magic weapon for things that "should", but you cannot be sure that they "are", unless you get c++ contracts.
Note that there is **no** run time overhead in release builds - they don't generate *any* code.

Now, imagine the following:
A user comes to the forum and tells you that `xy` yields wrong results when they do A, but with B it works. It doesn't crash, the results are just wrong, and nobody knows where the error happens.
Those are the bugs that take long to track down.
You ask for the code, run it *in your debug build that you should always have around*, the assert fails  and you tell the user what's wrong or fix the bug.

Long story short:
The asserts should stay in, and please use them to assert all things where you go "hmm, this *should* be OK if people (or you) used it correctly".
For things that are not performance critical, it's of course even better to validate inputs before you start a fit etc.

File: roofit/roofitcore/src/RooDataHist.cxx [hunk 2]
```diff
@@ -1007,16 +1009,16 @@ std::size_t RooDataHist::calcTreeIndex(const RooAbsCollection& coords, bool fast
       // Variable is not in external coordinates. Use current internal value.
       theVar = internalVar;
     }
-    // If fast is on, users promise that the sets have the same layout
-    assert(!fast || strcmp(internalVar->GetName(), theVar->GetName()) == 0);
+    // If fast is on, users promise that the sets have the same layout.
+    // assert(!fast || strcmp(internalVar->GetName(), theVar->GetName()) == 0);
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooDataHist.cxx:1013]: Same here and below.

**Review by hageboeck — CHANGES_REQUESTED** [review]
Nice to see that you find some opportunities for speed improvements!

File: roofit/roofitcore/src/RooDataHist.cxx (hunk 1 above)
Comment (hageboeck) [reply roofit/roofitcore/src/RooDataHist.cxx:993]: > For things that are not performance critical, it's of course even better to validate inputs before you start a fit etc.

I mean run-time checks that are active also in release builds, but this is mostly for the initialisation phase, while the asserts are better in the execution phase when you want speed.

**Comment by guitargeek** [comment]
Okay, I can update the code with less `auto`! And thanks for the heads up with the streamer.

**Comment by guitargeek** [comment]
Hi! Some updates on this PR:

1. I have made some of the `auto` explicit as suggested by @hageboeck 
2. @pcanal confirmed in private chat that filling a temporary `RooArgSet` is the right solution for the [legacy streamer](https://github.com/root-project/root/pull/7859/files#diff-f05bd2ea68e236a9a894f2e2575ec881427d17ce2d1211f6c830771afed03b68R2081)
3. I tested that RooDataHists written with ROOT 6.24 can still be read with this PR, as suggested by @lmoneta

File: roofit/roofitcore/inc/RooDataHist.h [hunk 3]
```diff
@@ -250,7 +251,7 @@ class RooDataHist : public RooAbsData, public RooDirItem {
   mutable double* _sumw2{nullptr}; //[_arrSize] Sum of weights^2
   double*         _binv {nullptr}; //[_arrSize] Bin volume array
 
-  RooArgSet  _realVars ; // Real dimensions of the dataset 
+  RooArgSet  _realVars ; // Real dimensions of the dataset
```
Comment (hageboeck) [inline roofit/roofitcore/inc/RooDataHist.h]: Try to keep formatting changes and code changes separate. (No need to change anything in the commit, though. 👍 )

File: roofit/roofitcore/src/RooDataHist.cxx [hunk 4]
```diff
@@ -1076,6 +1076,36 @@ RooPlot *RooDataHist::plotOn(RooPlot *frame, PlotOpt o) const
   return RooAbsData::plotOn(frame,o) ;
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// A faster version of RooDataHist::weight that assumes the passed arguments
+/// are aligned with the histogram variables.
+///
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooDataHist.cxx:1071]: Whenever you touch a comment, try to write full doxygen style with documentation of the parameters.
Random example:
https://github.com/root-project/root/blob/bd49faeb7952530b10073df939e329d92ceade95/roofit/roofitcore/src/RooBinSamplingPdf.cxx#L122-L126

I know it's more work, but if you don't do it right now, it won't get done in a long time.

File: roofit/roofitcore/src/RooHistPdf.cxx [hunk 5]
```diff
@@ -217,7 +217,7 @@ Double_t RooHistPdf::evaluate() const
     }
   }
 
-  Double_t ret = _dataHist->weight(_histObsList, _intOrder, !_unitNorm, _cdfBoundaries);
+  Double_t ret = _dataHist->weightFast(_histObsList, _intOrder, !_unitNorm, _cdfBoundaries);
 //  cout << "RooHistPdf::evaluate(" << GetName() << ") ret = " << ret << " ";
 //  cout << _histObsList[0] << " ";
 //  _histObsList[0]->Print("");
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooHistPdf.cxx:223]: Kill useless code when you get close to it.

File: roofit/roofitcore/inc/RooDataHist.h [hunk 6]
```diff
@@ -200,6 +200,13 @@ class RooDataHist : public RooAbsData, public RooDirItem {
   ///@}
   ////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
+  struct VarInfo {
```
Comment (hageboeck) [inline roofit/roofitcore/inc/RooDataHist.h:203]: Maybe one-line doxygen what this is.

File: roofit/roofitcore/src/RooDataHist.cxx [hunk 7]
```diff
@@ -2139,3 +2131,40 @@ void RooDataHist::getBatches(RooBatchCompute::RunContext& evalData, std::size_t
 void RooDataHist::registerWeightArraysToDataStore() const {
   _dstore->setExternalWeightArray(_wgt, _errLo, _errHi, _sumw2);
 }
+
+
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooDataHist.cxx:2131]: Also a bit of doxygen if you can.

File: roofit/roofitcore/src/RooDataHist.cxx [hunk 8]
```diff
@@ -1128,51 +1128,58 @@ Double_t RooDataHist::weight(const RooArgSet& bin, Int_t intOrder, Bool_t correc
   // Handle all interpolation cases
   _vars.assignValueOnly(bin) ;
 
-  auto varInfo = getVarInfo();
+  return weightInterpolated(_vars, intOrder, correctForBinSize, cdfBoundaries);
+}
+
+
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooDataHist.cxx:1134]: Doxygen.

File: roofit/roofitcore/src/RooDataHist.cxx [hunk 9]
```diff
@@ -1128,51 +1128,58 @@ Double_t RooDataHist::weight(const RooArgSet& bin, Int_t intOrder, Bool_t correc
...
+double RooDataHist::weightInterpolated(const RooArgSet& bin, int intOrder, bool correctForBinSize, bool cdfBoundaries) {
+  VarInfo const& varInfo = getVarInfo();
 
-  Double_t wInt(0) ;
+  auto centralIdx = calcTreeIndex(bin, true);
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooDataHist.cxx:1138]: That's very optional, but you could make temporary stuff `const`. You don't inadvertently overwrite it and make sure that it gets optimised out. (It will of course also be optimised out when it's not `const`, unless you write to it by accident.)

File: roofit/roofitcore/src/RooDataHist.cxx [hunk 10]
```diff
@@ -1266,67 +1271,58 @@ void RooDataHist::weightError(Double_t& lo, Double_t& hi, ErrorType etype) const
 /// Perform boundary safe 'intOrder'-th interpolation of weights in dimension 'dim'
 /// at current value 'xval'
```
Comment (hageboeck) [inline roofit/roofitcore/src/RooDataHist.cxx:1273]: If you could add the documentation of the parameters here? Now is probably the best time, because you had to look at what they mean.

File: roofit/roofitcore/test/testRooDataHist.cxx [hunk 11]
```diff
@@ -629,3 +631,64 @@ TEST(RooDataHist, AnalyticalIntegration)
...
+      hist.Fill(1.5, 0.5);
+   }
+   for(int i = 0; i < a1; ++i) {
+      hist.Fill(0.5, 1.5);
+   }
+   for(int i = 0; i < b1; ++i) {
+      hist.Fill(1.5, 1.5);
+   }
+
+   hist.Print();
```
Comment (hageboeck) [inline roofit/roofitcore/test/testRooDataHist.cxx:658]: We try to be reasonably silent in unit tests.

File: roofit/roofitcore/test/testRooDataHist.cxx [hunk 12]
```diff
@@ -629,3 +631,64 @@ TEST(RooDataHist, AnalyticalIntegration)
...
+
+   RooRealVar x("x", "x", 0, 2);
+   RooRealVar y("y", "y", 0, 2);
+
+   RooDataHist dataHist("data", "dataset with (x,y)", RooArgList(x, y), &hist);
+
+   dataHist.Print();
```
Comment (hageboeck) [inline roofit/roofitcore/test/testRooDataHist.cxx:665]: Also silence this if possible.

**Review by hageboeck — COMMENTED** [review]
It's looking good, but I have some suggestions to take it all the way.

Sorry that I insist on documentation, but you will never be so close to getting the docs done, because you should still remember what the functions and parameters mean.
In 3 months, this documentation won't get touched again.

Where is the schema evolution read test that you mentioned? Is it one of the tests that I implemented?

File: roofit/roofitcore/test/testRooDataHist.cxx (hunk 12 above)
Comment (guitargeek) [reply roofit/roofitcore/test/testRooDataHist.cxx:665]: Okay, I'll delete that print statement!

File: roofit/roofitcore/inc/RooDataHist.h (hunk 6 above)
Comment (guitargeek) [reply roofit/roofitcore/inc/RooDataHist.h:203]: Okay!

File: roofit/roofitcore/src/RooDataHist.cxx (hunk 10 above)
Comment (guitargeek) [reply roofit/roofitcore/src/RooDataHist.cxx:1273]: Okay!

File: roofit/roofitcore/src/RooDataHist.cxx (hunk 9 above)
Comment (guitargeek) [reply roofit/roofitcore/src/RooDataHist.cxx:1138]: Sure, that's a good thing to do!

File: roofit/roofitcore/src/RooDataHist.cxx (hunk 7 above)
Comment (guitargeek) [reply roofit/roofitcore/src/RooDataHist.cxx:2131]: Done.

File: roofit/roofitcore/src/RooHistPdf.cxx (hunk 5 above)
Comment (guitargeek) [reply roofit/roofitcore/src/RooHistPdf.cxx:223]: Good idea

File: roofit/roofitcore/src/RooDataHist.cxx (hunk 4 above)
Comment (guitargeek) [reply roofit/roofitcore/src/RooDataHist.cxx:1071]: Yes, I can do that.

File: roofit/roofitcore/src/RooDataHist.cxx (hunk 8 above)
Comment (guitargeek) [reply roofit/roofitcore/src/RooDataHist.cxx:1134]: Okay, added.

**Comment by guitargeek** [comment]
Hi @hageboeck! For the schema evolution from the previous version, I just made a test myself where I wrote a RooDataHist to a file with ROOT 6.24 and read it back with ROOT master + this PR. Do you think I should implement a unit test for that? I didn't think this was necessary, after all the change is rather trivial (removal of `_realVars` class member).

**Review by hageboeck — APPROVED** [review]
Very nice! 👍
